### PR TITLE
DEV: Fix meson debian python build

### LINKED ISSUE #16054: Bug: Meson build with dev.py fails to detect SciPy with debian python
While trying to build SciPy with meson on a 32bit machine for [#15814](https://github.com/scipy/scipy/issues/15814)  (using docker image `ubuntu:i386:bionic`; though this is not an architecture specific bug, I think)

After installing python3.8  and other required deps (through `apt` in ubuntu), and then python deps with pip3 (installed through [get-pip.py](https://bootstrap.pypa.io/get-pip.py)).

On running `python3.8 dev.py -j 2 --build-only` , build goes through fine; even the installation goes through fine.
```
Meson build setup OK
Build OK
Installing, see meson-install.log...
Installation OK
```

But later the script fails when [`get_project_info`](https://github.com/scipy/scipy/blob/653fd090fe18e28254986cd4c446c47b7051ef57/dev.py#L385) is called in dev.py script to import scipy, with the error message attached below.

```bash
Trying to find scipy from development installed path at: /__w/scipy/scipy/build-install/lib/python3.8/site-packages
Traceback (most recent call last):
  File "dev.py", line 393, in get_project_info
    test, version, mod_path = runtests.import_module()
  File "/__w/scipy/scipy/runtests.py", line 356, in import_module
    __import__(PROJECT_MODULE)
ModuleNotFoundError: No module named 'scipy'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "dev.py", line 635, in <module>
    main(argv=sys.argv[1:])
  File "dev.py", line 282, in main
    test, version, mod_path = get_project_info()
  File "dev.py", line 403, in get_project_info
    test, version, mod_path = runtests.import_module()
  File "/__w/scipy/scipy/runtests.py", line 356, in import_module
    __import__(PROJECT_MODULE)
ModuleNotFoundError: No module named 'scipy'
Error: Process completed with exit code 1.
```

The problem is that in dev.py, [`get_site_packages`](https://github.com/scipy/scipy/blob/653fd090fe18e28254986cd4c446c47b7051ef57/dev.py#L556) method returns `'/scipy/build-install/lib/python3.8/site-packages'` but actually meson installs scipy to `'/scipy/build-install/lib/python3/dist-packages'`

Feel free to check logs about the error and for reproducibility [here](https://github.com/AnirudhDagar/scipy/runs/6197379647?check_suite_focus=true).

### Additional Information
When I run ` python3.8 -c "import site; print(site.getsitepackages())"` I get the following:
```
['/usr/local/lib/python3.8/dist-packages', '/usr/lib/python3/dist-packages', '/usr/lib/python3.8/dist-packages']
```

I guess this issue is also discussed in mesonbuild/meson#8739, explaining problems with debian python and dist-packages vs site-packages or python3.x vs python3, depending on what python (system, venv, compiled) is being used.

Edit 1: Adding more info for someone who reads this later. Great blog on [Python, Debian, and the install locations](https://ffy00.github.io/blog/02-python-debian-and-the-install-locations/) and meson patching similar problem in https://github.com/mesonbuild/meson/pull/9288 

#### SciPy/NumPy/Python version information
`1.9.0.dev0+1925.653fd09 1.19.0 sys.version_info(major=3, minor=8, micro=0, releaselevel='final', serial=0)`

### Pitch
I can suggest a hacky fix to tweak `dev.py` specifically in [`get_project_info`](https://github.com/scipy/scipy/blob/1a945cae6ddc9186318cf34cfe1756cdd86eea34/dev.py#L385) or more specifically `PATH_INSTALLED` variable to also fall back to check for other possible installation paths that meson used in debian like here, for example utilizing path from `['/usr/local/lib/python3.8/dist-packages', '/usr/lib/python3/dist-packages', '/usr/lib/python3.8/dist-packages']` and stripping them to
`['/lib/python3.8/dist-packages', '/lib/python3/dist-packages', '/lib/python3.8/dist-packages']` and then joining it with the prefix.

There is probably a better way to do this. I can send a PR after a discussion if this is indeed a valid bug.

cc @rgommers @Smit-create

### PR DESCRIPTION
**Pull request by AnirudhDagar**
#### Reference issue
Closes gh-16054

#### What does this implement/fix?
This is a hacky solution for falling back from site-packages path dir to trying dist-packages path dir in case of debian python based meson build.

cc @rgommers

### CODE REVIEW

**Review by rgommers — COMMENTED** [review]
Thanks @AnirudhDagar! This LGTM for `dev.py` - we have the same issue with `do.py` though, so it'd be great if you could sync your fix to that file.

Sorry for the duplication, that's a very temporary state. I think we'll do `rm dev.py && mv do.py dev.py`` soon.

**Comment by AnirudhDagar** [comment]
I see, missed that! I'll take a look at and make the same changes in do.py.

**Review by schettino72 — CHANGES_REQUESTED** [review]
Thanks for taking the trouble to fix in 2 different implementations.

Regarding `do.py`, even before this problem was raised I noticed how messy dealing with paths in all different scenarios and systems is. 
That is why I created this `Dirs` class, to encapsulate the mess :grin:

So I would suggest that your changes modify only the `Dirs` class.
The `site` attribute could be made into a property that either gets site_packages or dist_packages. So the rest of the code should *not* be changed at all.
Makes sense? 

No where on task logic it needs to know if the site/installation path is debian specific or not.

**Comment by AnirudhDagar** [comment]
> That is why I created this Dirs class, to encapsulate the mess 😁
> So I would suggest that your changes modify only the Dirs class.
The site attribute could be made into a property that either gets site_packages or dist_packages. So the rest of the code should not be changed at all.
Makes sense?

Yes, indeed `Dirs` is a good way to clean up things, and I tried making the change there as the first thing too. But how do you decide if it gets site_packages or dist_packages? With the current logic, we check the existence of these directories:
Check for site_packages existence -> if not found -> check for dist_packages -> exists (ok so let's use this path).
But `Dirs` is independent of the build.

Consider these two separate cases:
1. you are building for the very first time. Hence there is no existence of any install dirs (build-install/.../site-packages or build-install/.../dist-packages). How do you choose to set the `self.site` attribute in `Dirs`? Ok, let's assume we do set it to site-packages all the time and then fallback later to debian in case it's missing. But that anyway fails for the second case.
2. you start building docs: This line needs to set the PYTHONPATH doesn't it? https://github.com/scipy/scipy/blob/f3ea971ba459ea640809bc70442b997b41338622/do.py#L920 How do you set it if you are running it for the very first time without the existence of any build/install?

Thanks for the review @schettino72, maybe I did not explain clearly or I'm probably missing something from your solution.

**Comment by FFY00** [comment]
> After reading the discussion in [mesonbuild/meson#9288](https://github.com/mesonbuild/meson/pull/9288) and [ev-br/mc_lib#45 (comment)](https://github.com/ev-br/mc_lib/issues/45#issuecomment-939519868), I'd like to propose a different/better approach (**the only problem is that it has to depend on `distutils`**, which I believe is fine because meson [depends](https://github.com/mesonbuild/meson/issues/4267#issuecomment-921383139) on that anyway; when debian folks finally start patching `sysconfig` instead of distutils which I hope they will before 3.12 then we can remove distutils here.

`distutils` will be removed in 3.12 and all of this will change, so depending on it isn't really a problem.

> To explain my point and problems involved, I'm commiting the changes that I made in `do.py` with the current approach. Also suggesting better approach below (adopted from logic in [mesonbuild/meson#9288](https://github.com/mesonbuild/meson/pull/9288)):

That logic is incomplete, I forgot about virtual environments, it also needs https://github.com/mesonbuild/meson/pull/9388.

**Comment by schettino72** [comment]
Oh ok. I missed that the code would be cleaned up with distutils.


Anyway, in case you guys decide to not include `distutils` my proposal was something like this:
note I did not run this code but the idea should be clear.

```python
@@ -296,10 +296,28 @@
             self.installed = Path(args.install_prefix).resolve()
         else:
             self.installed = self.build.parent / (self.build.stem + "-install")
-        # relative path for site-package with py version
-        # i.e. 'lib/python3.10/site-packages'
-        py_lib_path = Path(get_path('platlib')).relative_to(sys.exec_prefix)
-        self.site = self.installed / py_lib_path
+        self._site = None
+
+    @property
+    def site(self):
+        if self._site is None:
+            # relative path for site-package with py version
+            # i.e. 'lib/python3.10/site-packages'
+            py_lib_path = Path(get_path('platlib')).relative_to(sys.exec_prefix)
+            site = self.installed / py_lib_path
+            if site.exists():
+                self._site = site
+            else:
+                # Hack: fallback to debian based python dist-packages
+                # See https://github.com/scipy/scipy/issues/16054
+                py_version = f"python3.{sys.version_info[1]}"
+                dist_dir = Path(str(site).replace(py_version + "/site-packages",
+                                              "python3/dist-packages"))
+                if not dist_dir.exists():
+                    raise RuntimeError(f'Expected installation path "{site}" or '
+                                       f'"{dist_dir}" does not exist.')
+                self._site = dist_dir
+        return self._site
 
     def add_sys_path(self):
         """Add site dir to sys.path / PYTHONPATH"""
```

Making `site` a property you can delay the actual value to be set only when it is actually used (i.e. after the build step).
You would also need to amend the `doc` and `refguide-check` to delay using `Dirs.site`. 
You can pass a callable to CmdAction (as shown below) or they could be converted to a python-action like the `bench` task.

```python
from doit.action import CmdAction

    ...

    def create_cmd_string():
        return (f'cd doc; env PYTHONPATH="{dirs.site}" '
                f'make {" ".join(make_params)} {targets}'),

    ...
        'actions': [CmdAction(create_cmd_string)],
```

--------------------------------

Also note that while the name you used  `get_installed_path()` meaning `site-packages` or `dist-packages` makes sense. 
It is confusing because it clashes with the meaning of `Dirs.installed`. 
Here `Dirs.installed` path is the location used by `meson install` step as documented in the Dirs class docstring.

**Review by schettino72 — COMMENTED** [review]
You missed a reference to `dirs.site` on `copy_openblas()` line 530:

```
scipy_path = dirs.site / 'scipy'
```

This would not be a problem if there were no changes in the `Dirs` API :grinning:

**Comment by AnirudhDagar** [comment]
> Making site a property you can delay the actual value to be set only when it is actually used (i.e. after the build step).

Yes, that made sense to me.

> You would also need to amend the doc and refguide-check to delay using Dirs.site.

This is the part that I was more concerned about. I didn't know about the doit ` CmdAction ` which will solve the problem using a callable, thanks for pointing that out.

> You missed a reference to dirs.site on copy_openblas() line 530:
> This would not be a problem if there were no changes in the Dirs API 😀

It was intentionally left because `copy_openblas` is windows platform specific only.

> Also note that while the name you used get_installed_path() meaning site-packages or dist-packages makes sense.
> It is confusing because it clashes with the meaning of Dirs.installed.
> Here Dirs.installed path is the location used by meson install step as documented in the Dirs class docstring.

Yes, hence it was refactored to `install_prefix` in `dev.py`. See [commit](https://github.com/scipy/scipy/commit/0a6ef43b37040d7d2231ed6b2b506866454c6ff0) description, and `install_prefix` makes more sense anyway for that variable. I would have done the same for do.py but I just made the commits to share the code and how it was turning tedious compared to the distutils approach. I'm totally with you on keeping everything inside `Dirs` though :)

In any case, this is already becoming messy compared to the distutils based inference of path. I'll lean towards using the same logic (which also seems less hacky compared to checking the existence of installed path) that meson uses because, in the end, it is meson who is finding the install location. And yes, the debian virtual environments are not handled but that should be easy to handle there itself like @FFY00's https://github.com/mesonbuild/meson/pull/9388/.

**Comment by tylerjereddy** [comment]
I haven't forgotten to review this as discussed Wednesday--just recovering from a cold, I'll try to check this weekend.

**Comment by AnirudhDagar** [comment]
Thanks Tyler and no worries! Wish you a speedy recovery.

**Review by rgommers — COMMENTED** [review]
This looks reasonable now, and most of the changes are included in `Dirs`; what's outside is quite minor. Given that this solves the problem at hand, I think we can merge it.

@schettino72 I'm not sure I followed the `CmdAction` bit; how are you feeling about the current state of this PR?

**Comment by AnirudhDagar** [comment]
@rgommers please DON'T merge this in the current state. It is broken at the moment(specifically for building `doc` and `refguide-check`) and we have 2 possible fixes:

1. Using `distutils`, following exact meson logic to infer site-packages/dist-packages path for install path.
2. Infer install path in a hacky way(currently done in the PR), checking the existence of the directory as `build-install/.../site-packages` or `build-install/.../dist-packages`. Moving everything inside `Dirs` cleans up things and utilising the `CmdAction` to delay setting the path will be important here (explained below).

My question to the maintainers is which should we prefer?

---

To explain the second point in more detail, please consider the following possible workflow:
1. You are using a debian based system python. (Maybe in a docker container etc.)
2. You run `python do.py doc` to initiate automatically the following: build -> install -> doc build.

This will fail when running L921 because in L920 we set the PythonPath in the environment variable even before the meta-task `build` has been started. Doing something like this will not allow us to infer the PythonPath (`build-install/.../site-packages` or `build-install/.../dist-packages`) since it does not exist yet (because there is nothing installed yet, only after the build step we create the installed path):
https://github.com/scipy/scipy/blob/f3ea971ba459ea640809bc70442b997b41338622/do.py#L920-L921

To run the build `doc` or `refguide-check` commands, we need to export the path of SciPy for it to be visible (required in these commands in the terminal). Using `CmdAction` as suggested by @schettino72 will help solve that by actually delaying the setting of pythonpath (`cd doc; env PYTHONPATH="{dirs.site}`) after the meta-task `build` in this case and by that time we have an installation dir to infer the path from.

---

The other approach (point 1) is to just infer the python path exactly (without any guess work or waiting for the existence of any path) like how it is done in meson and there will be no need of making any hacks and a concise solution:

<details>
  <summary>Click to expand approach in point 1!</summary>
  
```python
# python3-distutils is assumed to be installed along side python3 in debian
# it is not part of the stdlib in patched debian python
import distutils.command.install  # somewhere at the top

def get_site_packages():
    """Depending on whether we have debian python or not, return site_dir path"""
    if 'deb_system' in distutils.command.install.INSTALL_SCHEMES:
        # debian patched python in use
        install_cmd = distutils.dist.Distribution().get_command_obj('install')
        install_cmd.select_scheme('deb_system')
        install_cmd.finalize_options()
        plat_path = Path(install_cmd.install_platlib)
    else:
        plat_path = Path(get_path('platlib'))
    return str(Path(PATH_INSTALLED) / plat_path.relative_to(sys.exec_prefix))
```
</details>


Sorry about this being slightly confusing, if you'd like I can probably just code up both the implementations and share them.

**Comment by schettino72** [comment]
`doit` has generally two phases. First in the "load" phase, it execute the `task_xxx()` functions to create task-metadata. And a `actions` execution phase... The problem here is that the actual path is not known during the load phase. It is known only after `build` is executed.

The solution is to delay getting the path information. Having a callable `CmdAction` would be one way to achieve this. Another one is using `getargs`, I will play with the code to see which one works better.

This would only be relevant if you decide not use `distutils`, as `distutils` can figure out which path will be used before the `build` task is executed.

**Comment by rgommers** [comment]
Thanks for the detailed write-up @AnirudhDagar, I get it now. It seems to me like the `distutils`-based approach is more concise and a little more robust (no try-except that could potentially trip up with multiple in-tree install dirs). So I think it's fine to rely on `distutils` for the next year. There is likely to be a replacement for that `INSTALL_SCHEMES` bit, because it's going to be needed in many places. So how about implementing that, and adding a comment above it saying _"If this needs to be replaced for Python 3.12 support and there's no stdlib alternative, use `CmdAction` as discussed in `gh-16058`"_?

**Comment by AnirudhDagar** [comment]
Cool, I'll update the PR with `distutils` based approach both for `dev.py` and `do.py`.

> So how about implementing that, and adding a comment above it saying "If this needs to be replaced for Python 3.12 support and there's no stdlib alternative, use CmdAction as discussed in gh-16058"?

I believe (mostly hope, can't say what they will do) that when 3.12 arrives, folks at Debian will move the patch from `distutils` to `sysconfig`. But for now, let's keep this comment linking the discussion.

**Review by schettino72 — APPROVED** [review]
nice. much cleaner :rocket:

**Review by rgommers — COMMENTED** [review]
Agreed, this looks great. One suggestion @AnirudhDagar, how about avoiding printing the deprecation warnings by using:
```python
with warnings.catch_warnings():
    warnings.filterwarnings("ignore", category=DeprecationWarning)
    from distutils import ....
```

**Comment by AnirudhDagar** [comment]
Yep, should have added that earlier to ignore the warning. This should be ready now!

**Review by rgommers — APPROVED** [review]
LGTM now, in it goes. Thanks @AnirudhDagar! And thanks @schettino72 for the reviews!

**Pull request merged**
